src/Page/Textile/Simulator.elm

```diff
@@ -416,6 +451,74 @@ update ({ textileDb, queries, navKey } as session) msg model =
                 |> updateQuery { query | yarnSize = yarnSize }
 
 
+toggleStepDetails : Int -> Maybe Int -> Maybe Int
+toggleStepDetails index detailedStep =
+    detailedStep
+        |> Maybe.map
+            (\current ->
+                if index == current then
+                    Nothing
+
+                else
+                    Just index
+            )
+        |> Maybe.withDefault (Just index)
+
+
+updateExistingMaterial : Inputs.Query -> Model -> Session -> Inputs.MaterialInput -> Material -> ( Model, Session, Cmd Msg )
+updateExistingMaterial query model session oldMaterial newMaterial =
+    let
+        materialQuery : Inputs.MaterialQuery
+        materialQuery =
+            { id = newMaterial.id
+            , share = oldMaterial.share
+            , spinning = Nothing
+            }
+    in
+    model
+        |> update session (SetModal NoModal)
+        |> updateQuery (Inputs.updateMaterial oldMaterial.material.id materialQuery query)
+
+
+updateMaterial : Inputs.Query -> Model -> Session -> Maybe Inputs.MaterialInput -> Autocomplete Material -> ( Model, Session, Cmd Msg )
+updateMaterial query model session maybeOldMaterial autocompleteState =
+    Maybe.map2
+        (updateExistingMaterial query model session)
+        maybeOldMaterial
+        (Autocomplete.selectedValue autocompleteState)
+        |> Maybe.withDefault
+            -- Add a new Material
+            (model
+                |> update session (SetModal NoModal)
+                |> selectMaterial autocompleteState
+            )
+
+
+selectMaterial : Autocomplete Material -> ( Model, Session, Cmd Msg ) -> ( Model, Session, Cmd Msg )
+selectMaterial autocompleteState ( model, session, commands ) =
```

Comment:
Putting this here but it's probably not in this specific code block that it should be done: when testing the UI, I realize when selecting an Element using the keyboard (arrow keys then pressing Enter), the modal closes and the focus is then positioned on the "add bookmark button" in the sidebar, which makes using the app using the keyboard difficult (you have to switch to using the mouse).

What would be great is to set the focus on the quantity input after an element is selected and the autocomplete modal closes, so we can immediately define it's 50% or 200g, then tab again to gain focus on the "add new element" btn, rinse and repeat. Full keyboard navigation.

I have a feeling it's easy to implement in the update function when an element is selected but I may be wrong; if it's trivial, let's impolement it in this PR, if it's not, let's create a card because I think this would add significant UX value to the app.